pkg/util/limiter/memory_limiter.go

```diff
@@ -0,0 +1,379 @@
+// Red - The service is undergoing severe memory pressure. All except the first job will be paused (above the redThreshold).
+type State int
+
+const (
+	Green  State = 0
+	Yellow State = 1
+	Red    State = 2
+)
+
+// HeapMemoryLimiter is the implementation of MemoryLimiter that watches Heap allocation to detect memory pressure.
+type HeapMemoryLimiter struct {
+	failFast bool // failFast specifies whether the jobs should fail immediately instead of waiting for the desired state.
+
+	redThreshold    uint64 // threshold in bytes above which the the system enters red state.
+	yellowThreshold uint64 // threshold in bytes above which the system enters yellow state. below this the system will be in green state.
+
+	yellowCond *cond // used for broadcasting when the state has changed to yellow.
+	greenCond  *cond // used for broadcasting when the state has changed to green.
+
+	jobsQueue []JobID    // FIFO queue for jobs.
+	jobsMutex sync.Mutex // mutex for accessing the queue.
+
+	state      State      // keeps track of the current status.
+	stateMutex sync.Mutex // mutex for accessing state
+
+	// memCheckInterval is the interval to read MemStats.
+	memCheckInterval time.Duration
+	ticker           *time.Ticker
+
+	// readMemStatsFn is the function to read the mem values is set as a reference to help with
+	// testing different values.
+	readMemStatsFn func(m *runtime.MemStats)
+
+	logger     log.Logger
+	stateGuage prometheus.Gauge
+}
+
+func (ml *HeapMemoryLimiter) NewJob() Job {
+	id := time.Now().UnixNano()
+	j := &job{
+		failFast: ml.failFast,
+		id:       JobID(id),
+		limiter:  ml,
+	}
+	ml.enqueue(j)
+	return j
+}
+
+// enqueue adds a job to the queue.
+func (ml *HeapMemoryLimiter) enqueue(j Job) {
+	ml.jobsMutex.Lock()
+	defer ml.jobsMutex.Unlock()
+	ml.jobsQueue = append(ml.jobsQueue, j.ID())
+}
+
+// dequeue removes a job from the queue.
+func (ml *HeapMemoryLimiter) dequeue(j Job) {
+	ml.jobsMutex.Lock()
+	defer ml.jobsMutex.Unlock()
+	for i, v := range ml.jobsQueue {
+		if v == j.ID() {
+			ml.jobsQueue = append(ml.jobsQueue[:i], ml.jobsQueue[i+1:]...)
+			break
+		}
+	}
+}
+
+// Wait provides a way for goroutines to slow down if there is memory pressure.
+// If memory pressure is detected, wait will block until the memory pressure is mitigated or the context is done.
+// If allowFirst is true, the first request in the queue will be always allowed.
+func (ml *HeapMemoryLimiter) Wait(ctx context.Context, id JobID, desiredState State, allowFirst bool, failFast bool) error {
+	if allowFirst && ml.isFirstJob(id) {
+		level.Debug(ml.logger).Log("msg", "Allowing the first job in the queue.")
+		// Allow the first job in the queue to continue.
+		return nil
+	}
+
+	ml.stateMutex.Lock()
+	currentState := ml.state
+	ml.stateMutex.Unlock()
+
+	if currentState <= desiredState {
+		// Allow if the desired state is already achieved
+		return nil
+	}
+
+	if failFast {
+		// If running in failFast mode, fail instead of waiting.
```

Comment:
I am thinking if we still need fail fast mode if we have the max wait time. What about also applying max wait time here also? 
We wait for some time before failing the query. For me it is different than waiting for some time and continue the query.